### PR TITLE
[cuda] Implement HAL semaphore using CUevent objects

### LINKED ISSUE #4727: Implement HAL semaphore using CUDA events
IREE uses semaphores to synchronize host<->device and device<->device across command buffers. The only efficient way to synchronize different streams in cuda  is to use CUEvents. We need to find a way to emulate HAL semaphore using CUevent.

### PR DESCRIPTION
**Pull request by antiagainst**
This commit adds a HAL semaphore implementation for the CUDA driver backed by iree_event_t and CUevent objects for different synchronization directions.

Fixes https://github.com/openxla/iree/issues/4727
Progress towards https://github.com/openxla/iree/issues/13245

### CODE REVIEW

File: experimental/cuda2/README.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+We need to fill the remaining capability gaps. Before going into details,
+the overall approach would be to:
+
+* State: we need a 64-bit integer value timeline. Given the binary state of
+  a `CUevent`, each `CUevent` would just be a "timepoint" on the timeline.
+* Ordering: we need to defer releasing the workload to the GPU until the
+  semaphore waits are reached on the host, or we can have some device
+  `CUevent` to wait on.
+* Direction: host to host and host to device is mssing; we can support that
+  with host synchronization mechanisms.
```
Comment (ScottTodd) [inline experimental/cuda2/README.md]:
```suggestion
* Direction: host to host and host to device is missing; we can support that
  with host synchronization mechanisms.
```

File: experimental/cuda2/README.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+##### GPU signal
+
+GPU signals can only be through a `CUevent` object, which has a binary state.
+We need to advance the timeline too. One way is to use `cuLaunchHostFunc()`
+to perform the advance from the CPU side. This additionally would mean we can
+reuse the logic form CPU signal to unblock CPU waits.
+
+For GPU waits, we can also leverage the same logic for CPU signals and deferred
+GPU queue actions. Though this is performant, given we are involving the CPU
+for GPU internal synchronization. We want to use `CUevent` for that:
```
Comment (ScottTodd) [inline experimental/cuda2/README.md]: I'm not following the grammar here. Can you rephrase?

File: experimental/cuda2/README.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+* We may not see GPU signal before seeing GPU wait requests, then we can also
+  keep track of all GPU waits in the timeline. Later once see either a CPU
+  signal or GPU signal advancing past the waited value, we can handle them
+  accordingly--submitting immediately or associating the `CUevent`.
+  This would also guarantee the requirement of `CUevent`--recording should
+  happen before waiting.
+* We can use the same `CUevent` to unblock multiple GPU waits; that's allowed.
+  Though it would mean we need to be careful regarding `CUevent` lifetime
+  management. Here we can use reference counting to see how many timepoints
+  are using it. And automatically return to a pool once done.
```
Comment (ScottTodd) [inline experimental/cuda2/README.md]: Grammer nits

```suggestion
* We can use the same `CUevent` to unblock multiple GPU waits. That's allowed,
  though it would mean we need to be careful regarding `CUevent` lifetime
  management. Here we can use reference counting to see how many timepoints
  are using it and automatically return to a pool once done.
```

File: experimental/cuda2/README.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  acquires CPU/GPU event objects there.
+* `iree_hal_cuda_timeline_semaphore_t`: contains a list of CPU wait and GPU
+  wait/signal timepoints.
+* `iree_hal_cuda2_queue_action_t`: a pending queue action (kernel launch or
+  stream-ordered allocation).
+* `iree_hal_cuda2_pending_queue_actions_t`: a data structure to manage pending
+  queue actions. It provides APIs to enqueue actions, and dvance the queue on
+  demand--queue actions are released to the GPU when all their wait semaphores
+  are signaled past the desired value, or we can have a `CUevent` object already
+  recorded to some `CUstream` to wait on.
```
Comment (ScottTodd) [inline experimental/cuda2/README.md]:
```suggestion
* `iree_hal_cuda2_pending_queue_actions_t`: a data structure to manage pending
  queue actions. It provides APIs to enqueue actions, and advance the queue on
  demand--queue actions are released to the GPU when all their wait semaphores
  are signaled past the desired value, or we can have a `CUevent` object already
  recorded to some `CUstream` to wait on.
```

File: experimental/cuda2/event_pool.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+CUevent iree_hal_cuda2_event_handle(const iree_hal_cuda2_event_t* event);
+
+// Retains the given |event| by increasing its reference count.
+void iree_hal_cuda2_event_retain(iree_hal_cuda2_event_t* event);
+
+// Releases the given |event| by decreasing its reference count.
+//
+// Afterwards releasing, if no other outstanding owner, |event| will be returned
+// to its owning pool.
+void iree_hal_cuda2_event_release(iree_hal_cuda2_event_t* event);
```
Comment (ScottTodd) [inline experimental/cuda2/event_pool.h]: nit: rephrase / grammar

```suggestion
// Releases the given |event| by decreasing its reference count.
//
// |event| will be returned to its owning pool when the reference count is 0.
void iree_hal_cuda2_event_release(iree_hal_cuda2_event_t* event);
```

File: experimental/cuda2/event_pool.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+// A simple pool of iree_hal_event_t objects to recycle.
+//
+// Thread-safe; multiple threads may acquire and release events from the pool.
+typedef struct iree_hal_cuda2_event_pool_t iree_hal_cuda2_event_pool_t;
+
+// Allocates a new event pool with up to |available_capacity| events.
+//
+// Extra events requests beyond the capability are directly created and
+// destroyed without pooling.
+iree_status_t iree_hal_cuda2_event_pool_allocate(
```
Comment (ScottTodd) [inline experimental/cuda2/event_pool.h]:
```suggestion
// Allocates a new event pool with up to |available_capacity| events.
//
// Extra events requested beyond the capability are directly created and
// destroyed without pooling.
iree_status_t iree_hal_cuda2_event_pool_allocate(
```

File: experimental/cuda2/cts/CMakeLists.txt [hunk 7]
```diff
@@ -20,8 +20,6 @@ iree_hal_cts_test_suite(
   EXCLUDED_TESTS
     # HAL event is unimplemented for now.
     "event"
-    # HAL semaphore is in the process of being implemented.
-    "semaphore"
```
Comment (ScottTodd) [inline experimental/cuda2/cts/CMakeLists.txt:24]: Want to add any more test cases to https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/cts/semaphore_test.h ? We could add more device signal/wait coverage...

File: experimental/cuda2/timepoint_pool.h [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    iree_event_pool_t* host_event_pool,
+    iree_hal_cuda2_event_pool_t* device_event_pool,
+    iree_host_size_t available_capacity, iree_allocator_t host_allocator,
+    iree_hal_cuda2_timepoint_pool_t** out_timepoint_pool);
+
+// Deallocates an timepoint pool and destroys all timepoints.
+//
+// All timepoints that were acquired from the pool must have already been
+// released back to it prior to deallocation.
+void iree_hal_cuda2_timepoint_pool_free(
```
Comment (ScottTodd) [inline experimental/cuda2/timepoint_pool.h]:
```suggestion
// Deallocates a timepoint pool and destroys all timepoints.
//
// All timepoints that were acquired from the pool must have already been
// released back to it prior to deallocation.
void iree_hal_cuda2_timepoint_pool_free(
```

File: experimental/cuda2/timepoint_pool.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    iree_hal_cuda2_timepoint_pool_t* timepoint_pool);
+
+// Acquires one or more timepoints from the timepoint pool.
+//
+// |out_timepoints| are owned by the caller and must be kept live until the
+// timepoints have been reached, or cancelled by the caller.
+iree_status_t iree_hal_cuda2_timepoint_pool_acquire_host_wait(
+    iree_hal_cuda2_timepoint_pool_t* timepoint_pool,
+    iree_host_size_t timepoint_count,
+    iree_hal_cuda2_timepoint_t** out_timepoints);
```
Comment (ScottTodd) [inline experimental/cuda2/timepoint_pool.h:100]: I only see these functions called with `timepoint_count == 1`. Is it useful to support multiple? Could reduce complexity a bit by making these just take a single `out_timepoint` to be acquired.

File: experimental/cuda2/timepoint_pool.c [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+    memcpy(out_timepoints, &timepoint_pool->available_list[pool_base_index],
+           from_pool_count * sizeof(*timepoint_pool->available_list));
+    timepoint_pool->available_count -= from_pool_count;
+    remaining_count -= from_pool_count;
+  }
+  iree_slim_mutex_unlock(&timepoint_pool->timepoint_mutex);
+
+  // Allocate the rest of the timepoints.
+  if (remaining_count > 0) {
+    IREE_TRACE_ZONE_BEGIN(z0);
```
Comment (ScottTodd) [inline experimental/cuda2/timepoint_pool.c]: Why start the zone down here? Seems more accurate to begin the zone before `iree_slim_mutex_lock`.

File: experimental/cuda2/timepoint_pool.c [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+      z0, iree_hal_cuda2_event_pool_acquire(timepoint_pool->device_event_pool,
+                                            timepoint_count, device_events));
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_cuda2_timepoint_pool_acquire_internal(
+              timepoint_pool, timepoint_count, out_timepoints));
+  for (iree_host_size_t i = 0; i < timepoint_count; ++i) {
+    out_timepoints[i]->kind = IREE_HAL_CUDA_TIMEPOINT_KIND_DEVICE_SIGNAL;
+    out_timepoints[i]->timepoint.device_signal = device_events[i];
+  }
```
Comment (ScottTodd) [inline experimental/cuda2/timepoint_pool.c:261]: Should this release `device_events` with `iree_hal_cuda2_event_release` if `iree_hal_cuda2_timepoint_pool_acquire_internal` fails? Could switch to `if (iree_status_is_ok(status))` error handling style and add cleanup or assignment code at the bottom of this function.

Same question for the following function (`iree_hal_cuda2_timepoint_pool_acquire_device_wait`). Although... if these functions are only different in which `IREE_HAL_CUDA_TIMEPOINT_KIND_*` value they set (and whether they set `.device_wait` or `.device_signal`, they could both go through a helper function instead.

---

If adding a helper function, you could keep the trace zones as-is with

```c
iree_status_t iree_hal_cuda2_timepoint_pool_acquire_device_signal() {
  IREE_TRACE_ZONE_BEGIN(z0);
  iree_status_t status = helper_function(..., KIND_DEVICE_SIGNAL);
  IREE_TRACE_ZONE_END(z0);
  return status;
}

iree_status_t iree_hal_cuda2_timepoint_pool_acquire_device_wait() {
  IREE_TRACE_ZONE_BEGIN(z0);
  iree_status_t status = helper_function(..., KIND_DEVICE_WAIT);
  IREE_TRACE_ZONE_END(z0);
  return status;
}

iree_status_t helper_function(..., kind) {
  IREE_TRACE_ZONE_BEGIN(z0);  // optional, maybe omit
  // ...
  IREE_TRACE_ZONE_END(z0);
}
```

File: experimental/cuda2/timepoint_pool.c [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+           to_pool_count * sizeof(*timepoint_pool->available_list));
+    timepoint_pool->available_count += to_pool_count;
+    remaining_count -= to_pool_count;
+  }
+  iree_slim_mutex_unlock(&timepoint_pool->timepoint_mutex);
+
+  // Deallocate the rest of the timepoints. We don't bother resetting them as we
+  // are getting rid of them.
+  if (remaining_count > 0) {
+    IREE_TRACE_ZONE_BEGIN(z0);
```
Comment (ScottTodd) [inline experimental/cuda2/timepoint_pool.c]: (Similar to another comment) Why start the zone so far down in the function? Seems like this should include the mutex.

I looked around to see if the tracing macros weren't thread-safe in some way, and there are uses matching what I'd expect, e.g. https://github.com/openxla/iree/blob/2d053fd69d847dd34e89474a09f73d896e2f3820/runtime/src/iree/hal/driver_registry.c#L89-L93

File: experimental/cuda2/event_pool.c [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+//===----------------------------------------------------------------------===//
+
+struct iree_hal_cuda2_event_t {
+  // The allocator used to create the event.
+  iree_allocator_t host_allocator;
+  // The symbols used to create and destroy CUevent objects.
+  const iree_hal_cuda2_dynamic_symbols_t* symbols;
+
+  // The event pool that owns this event.
+  iree_hal_cuda2_event_pool_t* pool;
```
Comment (ScottTodd) [inline experimental/cuda2/event_pool.c]: Do all events belong to an event pool, or can this be NULL? (I'm a bit confused by the ambiguity in what ref count values mean, and answering this would help clear that up)

File: experimental/cuda2/event_pool.c [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+    memcpy(out_events, &event_pool->available_list[pool_base_index],
+           from_pool_count * sizeof(*event_pool->available_list));
+    event_pool->available_count -= from_pool_count;
+    remaining_count -= from_pool_count;
+  }
+  iree_slim_mutex_unlock(&event_pool->event_mutex);
+
+  // Allocate the rest of the events.
+  if (remaining_count > 0) {
+    IREE_TRACE_ZONE_BEGIN(z0);
```
Comment (ScottTodd) [inline experimental/cuda2/event_pool.c]: Start trace zone higher?

File: experimental/cuda2/event_pool.c [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+           to_pool_count * sizeof(*event_pool->available_list));
+    event_pool->available_count += to_pool_count;
+    remaining_count -= to_pool_count;
+  }
+  iree_slim_mutex_unlock(&event_pool->event_mutex);
+
+  // Deallocate the rest of the events. We don't bother resetting them as we are
+  // getting rid of them.
+  if (remaining_count > 0) {
+    IREE_TRACE_ZONE_BEGIN(z0);
```
Comment (ScottTodd) [inline experimental/cuda2/event_pool.c]: Start trace zone higher?

File: experimental/cuda2/event_semaphore.c [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+  // Wait until the timepoint resolves.
+  // If satisfied the timepoint is automatically cleaned up and we are done. If
+  // the deadline is reached before satisfied then we have to clean it up.
+  status = iree_wait_one(&timepoint->timepoint.host_wait, deadline_ns);
+  if (!iree_status_is_ok(status)) {
+    iree_hal_semaphore_cancel_timepoint(&semaphore->base, &timepoint->base);
+  }
+  iree_hal_cuda2_timepoint_pool_release(semaphore->timepoint_pool, 1,
+                                        &timepoint);
+  return iree_ok_status();
```
Comment (ScottTodd) [inline experimental/cuda2/event_semaphore.c]: Should this return the status?

```suggestion
  return status;
```

File: experimental/cuda2/event_semaphore.c [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+      &semaphore->base, min_value, timeout,
+      (iree_hal_semaphore_callback_t){
+          .fn = iree_hal_cuda2_semaphore_timepoint_host_wait_callback,
+          .user_data = *out_timepoint,
+      },
+      &(*out_timepoint)->base);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_cuda2_semaphore_wait(
```
Comment (ScottTodd) [inline experimental/cuda2/event_semaphore.c:231]: Want to add tracing to this function?

Same for a few others in this file.

File: experimental/cuda2/event_pool.c (hunk 14 above)
Comment (antiagainst) [reply experimental/cuda2/event_pool.c]: This was mostly copied from `iree_event_pool_t`. I've inserted another level of tracing zone and make this one named--so that we can track both.

File: experimental/cuda2/timepoint_pool.c (hunk 10 above)
Comment (antiagainst) [reply experimental/cuda2/timepoint_pool.c]: This was mostly copied from `iree_event_pool_t`. I've changed to insert another level of tracing zone and make this one named--so that we can track both.

File: experimental/cuda2/event_pool.c (hunk 13 above)
Comment (antiagainst) [reply experimental/cuda2/event_pool.c]: Yes, as documented in `event_pool.h` we can only create these event objects from the event pool. I updated the comment to be clear.

File: experimental/cuda2/timepoint_pool.c (hunk 12 above)
Comment (antiagainst) [reply experimental/cuda2/timepoint_pool.c]: As explained in other places this is basically copying the way `iree_event_pool_t` is doing. I've updated it to two levels of trace zones.

File: experimental/cuda2/timepoint_pool.h (hunk 9 above)
Comment (antiagainst) [reply experimental/cuda2/timepoint_pool.h:100]: Probably. I wrote these functions in a bottom up way before having their usages done. The hope was that we can optimize by doing multi time point acquiring, given that avoids paying the locking cost multiple times. But it might not be really happening anytime soon. Though given this is done, I'd prefer just to keep it this way for now. :)

File: experimental/cuda2/timepoint_pool.c (hunk 11 above)
Comment (antiagainst) [reply experimental/cuda2/timepoint_pool.c:261]: > Should this release `device_events` with `iree_hal_cuda2_event_release` if `iree_hal_cuda2_timepoint_pool_acquire_internal` fails?

This is doing `iree_alloca` to grab stack memory space so it should be fine.



> if these functions are only different in which `IREE_HAL_CUDA_TIMEPOINT_KIND_*` value they set (and whether they set `.device_wait` or `.device_signal`, they could both go through a helper function instead.

Right now the majority of the logic is already in `iree_hal_cuda2_timepoint_pool_acquire_internal`. The alloca and post-update refer to different fields of the `union` too. I think the gain of trying to further unify is minor.

File: experimental/cuda2/cts/CMakeLists.txt (hunk 7 above)
Comment (antiagainst) [reply experimental/cuda2/cts/CMakeLists.txt:24]: Yes! I'd certainly love to. But with all the stuff on my plate I may need to come to this later. Sorry about it. :(

Comment (ScottTodd) [reply experimental/cuda2/cts/CMakeLists.txt:24]: Could you at least add TODOs for test cases that would be helpful to add, given your current context on this work?

**Review by ScottTodd — APPROVED** [review]

File: experimental/cuda2/api.h [hunk 18]
```diff
@@ -59,6 +59,13 @@ typedef struct iree_hal_cuda2_device_params_t {
   // transient allocations while also increasing memory consumption.
   iree_host_size_t arena_block_size;
 
+  // The host and device event pool capability.
+  // The CUDA driver implements semaphore with host and device events. This
+  // parameter controls the size of those pools. Larger values would make
+  // creating semaphore values quicker, though with increased memory
+  // consumption.
+  iree_host_size_t event_pool_capability;
```
Comment (benvanik) [inline experimental/cuda2/api.h]: s/capability/capacity/g

File: experimental/cuda2/pending_queue_actions.h [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    iree_hal_cuda2_pending_queue_actions_t;
+
+// Creates a pending actions queue.
+iree_status_t iree_hal_cuda2_pending_queue_actions_create(
+    const iree_hal_cuda2_dynamic_symbols_t* symbols,
+    iree_arena_block_pool_t* block_pool, iree_allocator_t host_allocator,
+    iree_hal_cuda2_pending_queue_actions_t** out_actions);
+
+// Destroys the pending actions queue.
+void iree_hal_cuda2_pending_queue_actions_destroy(iree_hal_resource_t* actions);
```
Comment (benvanik) [inline experimental/cuda2/pending_queue_actions.h]: create/destroy are used for reference counted stuff - allocate/free would be the terminology for something that is allocated/freed (vs retained/released)

File: experimental/cuda2/timepoint_pool.c [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+#include "iree/base/internal/event_pool.h"
+#include "iree/base/internal/synchronization.h"
+#include "iree/hal/api.h"
+#include "iree/hal/utils/semaphore_base.c"
+
+//===----------------------------------------------------------------------===//
+// iree_hal_cuda2_timepoint_t
+//===----------------------------------------------------------------------===//
+
+static iree_status_t iree_hal_cuda2_timepoint_create(
```
Comment (benvanik) [inline experimental/cuda2/timepoint_pool.c]: allocate/free terminology as there's no ref counting

**Review by benvanik — APPROVED** [review]
minor nits but otherwise LGTM! nice work!

File: experimental/cuda2/timepoint_pool.c (hunk 20 above)
Comment (antiagainst) [reply experimental/cuda2/timepoint_pool.c]: Got it. Done.

File: experimental/cuda2/pending_queue_actions.h (hunk 19 above)
Comment (antiagainst) [reply experimental/cuda2/pending_queue_actions.h]: Actually `iree_hal_cuda2_pending_queue_actions_t` is backed by `iree_hal_resource_t` so there is refcounting with it; and we retain it for each pending action to make sure it's not destructed before all actions inside are released.

File: experimental/cuda2/cts/CMakeLists.txt (hunk 7 above)
Comment (antiagainst) [reply experimental/cuda2/cts/CMakeLists.txt:24]: Yup! Added some potential test TODOs in https://github.com/openxla/iree/pull/14426/commits/6e65bfb0b0781819ec7c6bf5c4cd1d9de4eab638

Comment (antiagainst) [reply experimental/cuda2/cts/CMakeLists.txt:24]: actually https://github.com/openxla/iree/pull/14426/commits/43969d93038effa6d0305d4d40845d55286b7727
